## 1. The NIST Special Publication 800-63B
<p>If you – 50 years ago – needed to come up with a secret password you were probably part of a secret espionage organization or (more likely) you were pretending to be a spy when playing as a kid. Today, many of us are forced to come up with new passwords <em>all the time</em> when signing into sites and apps. As a password <em>inventeur</em> it is your responsibility to come up with good, hard-to-crack passwords. But it is also in the interest of sites and apps to make sure that you use good passwords. The problem is that it's really hard to define what makes a good password. However, <em>the National Institute of Standards and Technology</em> (NIST) knows what the second best thing is: To make sure you're at least not using a <em>bad</em> password. </p>
<p>In this notebook, we will go through the rules in <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a> which details what checks a <em>verifier</em> (what the NIST calls a second party responsible for storing and verifying passwords) should perform to make sure users don't pick bad passwords. We will go through the passwords of users from a fictional company and use python to flag the users with bad passwords. But us being able to do this already means the fictional company is breaking one of the rules of 800-63B:</p>
<blockquote>
  <p>Verifiers SHALL store memorized secrets in a form that is resistant to offline attacks. Memorized secrets SHALL be salted and hashed using a suitable one-way key derivation function.</p>
</blockquote>
<p>That is, never save users' passwords in plaintext, always encrypt the passwords! Keeping this in mind for the next time we're building a password management system, let's load in the data.</p>
<p><em>Warning: The list of passwords and the fictional user database both contain <strong>real</strong> passwords leaked from <strong>real</strong> websites. These passwords have not been filtered in any way and include words that are explicit, derogatory and offensive.</em></p>

In [178]:
# Importing the pandas module
# ... YOUR CODE FOR TASK 1 ...
import pandas as pd
# Loading in datasets/users.csv 
users = pd.read_csv('datasets/users.csv')

# Printing out how many users we've got
# ... YOUR CODE FOR TASK 1 ...
print(users.shape)

# Taking a look at the 12 first users
# ... YOUR CODE FOR TASK 1 ...
users.head(12)

(982, 3)


id         user_name          password
0    1    vance.jennings          joobheco
1    2    consuelo.eaton        0869347314
2    3   mitchel.perkins        fabypotter
3    4    odessa.vaughan         aharney88
4    5    araceli.wilder        acecdn3000
5    6  shawn.harrington           5278049
6    7        evelyn.gay            master
7    8       noreen.hale            murphy
8    9       gladys.ward           lwsves2
9   10   brant.zimmerman  1190KAREN5572497
10  11     leanna.abbott          aivlys24
11  12   milford.hubbard           hubbard

## 2. Passwords should not be too short
<p>If we take a look at the first 12 users above we already see some bad passwords. But let's not get ahead of ourselves and start flagging passwords <em>manually</em>. What is the first thing we should check according to the NIST Special Publication 800-63B?</p>
<blockquote>
  <p>Verifiers SHALL require subscriber-chosen memorized secrets to be at least 8 characters in length.</p>
</blockquote>
<p>Ok, so the passwords of our users shouldn't be too short. Let's start by checking that!</p>

In [180]:
# Calculating the lengths of users' passwords
users['length'] = users['password'].str.len()

# Flagging the users with too short passwords
users['too_short'] = users['length']<8

# Counting and printing the number of users with too short passwords
# ... YOUR CODE FOR TASK 2 ...
print(sum(users['too_short']))

# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 2 ...
users.head(12)

376


id         user_name          password  length too_short
0    1    vance.jennings          joobheco       8     False
1    2    consuelo.eaton        0869347314      10     False
2    3   mitchel.perkins        fabypotter      10     False
3    4    odessa.vaughan         aharney88       9     False
4    5    araceli.wilder        acecdn3000      10     False
5    6  shawn.harrington           5278049       7      True
6    7        evelyn.gay            master       6      True
7    8       noreen.hale            murphy       6      True
8    9       gladys.ward           lwsves2       7      True
9   10   brant.zimmerman  1190KAREN5572497      16     False
10  11     leanna.abbott          aivlys24       8     False
11  12   milford.hubbard           hubbard       7      True

## 3.  Common passwords people use
<p>Already this simple rule flagged a couple of offenders among the first 12 users. Next up in Special Publication 800-63B is the rule that</p>
<blockquote>
  <p>verifiers SHALL compare the prospective secrets against a list that contains values known to be commonly-used, expected, or compromised.</p>
  <ul>
  <li>Passwords obtained from previous breach corpuses.</li>
  <li>Dictionary words.</li>
  <li>Repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</li>
  <li>Context-specific words, such as the name of the service, the username, and derivatives thereof.</li>
  </ul>
</blockquote>
<p>We're going to check these in order and start with <em>Passwords obtained from previous breach corpuses</em>, that is, websites where hackers have leaked all the users' passwords. As many websites don't follow the NIST guidelines and encrypt passwords there now exist large lists of the most popular passwords. Let's start by loading in the 10,000 most common passwords which I've taken from <a href="https://github.com/danielmiessler/SecLists/tree/master/Passwords">here</a>.</p>

In [182]:
# Reading in the top 10000 passwords
common_passwords = pd.read_csv('datasets/10_million_password_list_top_10000.txt', header=None, squeeze=True)

# Taking a look at the top 20
#  YOUR CODE FOR TASK 3 
common_passwords.head(20)

0        123456
1      password
2      12345678
3        qwerty
4     123456789
5         12345
6          1234
7        111111
8       1234567
9        dragon
10       123123
11     baseball
12       abc123
13     football
14       monkey
15      letmein
16       696969
17       shadow
18       master
19       666666
Name: 0, dtype: object

## 4.  Passwords should not be common passwords
<p>The list of passwords was ordered, with the most common passwords first, and so we shouldn't be surprised to see passwords like <code>123456</code> and <code>qwerty</code> above. As hackers also have access to this list of common passwords, it's important that none of our users use these passwords!</p>
<p>Let's flag all the passwords in our user database that are among the top 10,000 used passwords.</p>

In [184]:
# Flagging the users with passwords that are common passwords
users['common_password'] = users['password'].isin(common_passwords)

# Counting and printing the number of users using common passwords
# ... YOUR CODE FOR TASK 4 ...
print(sum(users['common_password']))
# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 4 ...
users.head(12)

129


id         user_name          password  length too_short common_password
0    1    vance.jennings          joobheco       8     False           False
1    2    consuelo.eaton        0869347314      10     False           False
2    3   mitchel.perkins        fabypotter      10     False           False
3    4    odessa.vaughan         aharney88       9     False           False
4    5    araceli.wilder        acecdn3000      10     False           False
5    6  shawn.harrington           5278049       7      True           False
6    7        evelyn.gay            master       6      True            True
7    8       noreen.hale            murphy       6      True            True
8    9       gladys.ward           lwsves2       7      True           False
9   10   brant.zimmerman  1190KAREN5572497      16     False           False
10  11     leanna.abbott          aivlys24       8     False           False
11  12   milford.hubbard           hubbard       7      True           False

## 5. Passwords should not be common words
<p>Ay ay ay! It turns out many of our users use common passwords, and of the first 12 users there are already two. However, as most common passwords also tend to be short, they were already flagged as being too short. What is the next thing we should check?</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains [...] dictionary words.</p>
</blockquote>
<p>This follows the same logic as before: It is easy for hackers to check users' passwords against common English words and therefore common English words make bad passwords. Let's check our users' passwords against the top 10,000 English words from <a href="https://github.com/first20hours/google-10000-english">Google's Trillion Word Corpus</a>.</p>

In [186]:
# Reading in a list of the 10000 most common words
words = pd.read_csv('datasets/google-10000-english.txt', header=None, squeeze=True)

# Flagging the users with passwords that are common words
users['common_word'] = users['password'].str.lower().isin(words)

# Counting and printing the number of users using common words as passwords
# ... YOUR CODE FOR TASK 5 ...
print(sum(users['common_word']))
# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 5 ...
users.head(12)

137


id         user_name          password  length too_short common_password  \
0    1    vance.jennings          joobheco       8     False           False   
1    2    consuelo.eaton        0869347314      10     False           False   
2    3   mitchel.perkins        fabypotter      10     False           False   
3    4    odessa.vaughan         aharney88       9     False           False   
4    5    araceli.wilder        acecdn3000      10     False           False   
5    6  shawn.harrington           5278049       7      True           False   
6    7        evelyn.gay            master       6      True            True   
7    8       noreen.hale            murphy       6      True            True   
8    9       gladys.ward           lwsves2       7      True           False   
9   10   brant.zimmerman  1190KAREN5572497      16     False           False   
10  11     leanna.abbott          aivlys24       8     False           False   
11  12   milford.hubbard           hubbard       7      True           False   

   common_word  
0        False  
1        False  
2        False  
3        False  
4        False  
5        False  
6         True  
7         True  
8        False  
9        False  
10       False  
11       False

## 6. Passwords should not be your name
<p>It turns out many of our passwords were common English words too! Next up on the NIST list:</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains [...] context-specific words, such as the name of the service, the username, and derivatives thereof.</p>
</blockquote>
<p>Ok, so there are many things we could check here. One thing to notice is that our users' usernames consist of their first names and last names separated by a dot. For now, let's just flag passwords that are the same as either a user's first or last name.</p>

In [188]:
# Extracting first and last names into their own columns
users['first_name'] = users['user_name'].str.extract(r'(^\w+)', expand=False)
users['last_name'] = users['user_name'].str.extract(r'(\w+$)', expand=False)

# Flagging the users with passwords that matches their names
users['uses_name'] = (users['first_name'].str.lower() == users['password']) | ((users['last_name']).str.lower()== users['password'])

# Counting and printing the number of users using names as passwords
# ... YOUR CODE FOR TASK 6 ...
print(sum(users['uses_name']))

# Taking a look at the 12 first rows
# ... YOUR CODE FOR TASK 6 ...
users[['user_name','first_name', 'last_name','password','uses_name']]

50


user_name first_name   last_name          password uses_name
0        vance.jennings      vance    jennings          joobheco     False
1        consuelo.eaton   consuelo       eaton        0869347314     False
2       mitchel.perkins    mitchel     perkins        fabypotter     False
3        odessa.vaughan     odessa     vaughan         aharney88     False
4        araceli.wilder    araceli      wilder        acecdn3000     False
5      shawn.harrington      shawn  harrington           5278049     False
6            evelyn.gay     evelyn         gay            master     False
7           noreen.hale     noreen        hale            murphy     False
8           gladys.ward     gladys        ward           lwsves2     False
9       brant.zimmerman      brant   zimmerman  1190KAREN5572497     False
10        leanna.abbott     leanna      abbott          aivlys24     False
11      milford.hubbard    milford     hubbard           hubbard      True
12            mamie.fox      mamie         fox        mitonguito     False
13        jamie.cochran      jamie     cochran            310356     False
14   nathaniel.robinson  nathaniel    robinson         angelmajo     False
15           lorrie.gay     lorrie         gay           oZ4k0QE     False
16         domingo.dyer    domingo        dyer           chelsea     False
17       martin.pacheco     martin     pacheco           zvc1939     False
18        shelby.massey     shelby      massey            nickgd     False
19      rosella.barrett    rosella     barrett        O2gv3LlcfG     False
20        karina.morton     karina      morton          dada4943     False
21      leticia.sanford    leticia     sanford          cocacola     False
22        jenny.woodard      jenny     woodard           woodard      True
23      brandie.webster    brandie     webster          sentry31     False
24       sabrina.suarez    sabrina      suarez        OTEL3Q0D8y     False
25         dianna.munoz     dianna       munoz             AJ9Da     False
26         julia.savage      julia      savage            ewokzs     False
27         loretta.bass    loretta        bass     WvNV1aKyFEcPe     False
28      joaquin.walters    joaquin     walters           YyGjz8E     False
29           rene.small       rene       small          toreze00     False
..                  ...        ...         ...               ...       ...
952    christa.morrison    christa    morrison          mercedes     False
953      clarence.britt   clarence       britt         28may1997     False
954     carmela.clayton    carmela     clayton     N2XTArGRVhKl5     False
955         royce.combs      royce       combs   Ct3EayTGuHs4Ic2     False
956        devon.holman      devon      holman           raiders     False
957       becky.hickman      becky     hickman          AQiCWRGL     False
958        deena.holmes      deena      holmes      9xQUdbKNhYsW     False
959       mark.chandler       mark    chandler         ye6491982     False
960       carmelo.byers    carmelo       byers            asdfgh     False
961  mohammed.carpenter   mohammed   carpenter       ujXSn2dZWhF     False
962      rico.valentine       rico   valentine        ap10172203     False
963     angel.jefferson      angel   jefferson          51183208     False
964      chrystal.burns   chrystal       burns            DILWYN     False
965        irma.vasquez       irma     vasquez            spider     False
966         taylor.kent     taylor        kent            summer     False
967       deloris.dixon    deloris       dixon             seeks     False
968        julian.gross     julian       gross          Passion!     False
969          joey.poole       joey       poole        lagrimason     False
970        noel.montoya       noel     montoya           colours     False
971       josef.hoffman      josef     hoffman      pharmacy2012     False
972       jorge.patrick      jorge     patrick       09196921342     False
973       rogelio.payne    rogelio     

## 7. Passwords should not be repetitive
<p>Milford Hubbard (user number 12 above), what where you thinking!? Ok, so the last thing we are going to check is a bit tricky:</p>
<blockquote>
  <p>verifiers SHALL compare the prospective secrets [so that they don't contain] repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</p>
</blockquote>
<p>This is tricky to check because what is <em>repetitive</em> is hard to define. Is <code>11111</code> repetitive? Yes! Is <code>12345</code> repetitive? Well, kind of. Is <code>13579</code> repetitive? Maybe not..? To check for <em>repetitiveness</em> can be arbitrarily complex, but here we're only going to do something simple. We're going to flag all passwords that contain 4 or more repeated characters.</p>

In [190]:
### Flagging the users with passwords with >= 4 repeats
users['too_many_repeats'] = users['password'].str.contains(r'(.)\1\1\1')

# Taking a look at the users with too many repeats
# ... YOUR CODE FOR TASK 7 ...
users[users['too_many_repeats']==True]

id         user_name  password  length too_short common_password  \
146  147       patti.dixon    555555       6      True            True   
572  573  cornelia.bradley    555555       6      True            True   
644  645       essie.lopez     11111       5      True            True   
798  799       charley.key    888888       6      True            True   
807  808   thurman.osborne  rinnnng0       8     False           False   
941  942    mitch.ferguson    aaaaaa       6      True            True   

    common_word first_name last_name uses_name too_many_repeats  
146       False      patti     dixon     False             True  
572       False   cornelia   bradley     False             True  
644       False      essie     lopez     False             True  
798       False    charley       key     False             True  
807       False    thurman   osborne     False             True  
941       False      mitch  ferguson     False             True

## 8. All together now!
<p>Now we have implemented all the basic tests for bad passwords suggested by NIST Special Publication 800-63B! What's left is just to flag all bad passwords and maybe to send these users an e-mail that strongly suggests they change their password.</p>

In [192]:
# Flagging all passwords that are bad
users['bad_password'] = (users['too_short'])|(users['common_password'])|(users['common_word'])|(users['uses_name'])|(users['too_many_repeats'])

# Counting and printing the number of bad passwords
# ... YOUR CODE FOR TASK 8 ...
print(sum(users['bad_password']))

# Looking at the first 25 bad passwords
# ... YOUR CODE FOR TASK 8 ...
users[users['bad_password']==True]['password'].head(25)

424


5       5278049
6        master
7        murphy
8       lwsves2
11      hubbard
13       310356
15      oZ4k0QE
16      chelsea
17      zvc1939
18       nickgd
21     cocacola
22      woodard
25        AJ9Da
26       ewokzs
28      YyGjz8E
30         reid
34      jOYZBs8
38      wwewwf1
43       225377
45       NdZ7E6
47        CQB3Z
48        diffo
51    123456789
52      y8uM7D6
56      mikeloo
Name: password, dtype: object

## 9. Otherwise, the password should be up to the user
<p>In this notebook, we've implemented the password checks recommended by the NIST Special Publication 800-63B. It's certainly possible to better implement these checks, for example, by using a longer list of common passwords. Also note that the NIST checks in no way guarantee that a chosen password is good, just that it's not obviously bad.</p>
<p>Apart from the checks we've implemented above the NIST is also clear with what password rules should <em>not</em> be imposed:</p>
<blockquote>
  <p>Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets. Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically).</p>
</blockquote>
<p>So the next time a website or app tells you to "include both a number, symbol and an upper and lower case character in your password" you should send them a copy of <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a>.</p>

In [194]:
# Enter a password that passes the NIST requirements
# PLEASE DO NOT USE AN EXISTING PASSWORD HERE
new_password = "test@1986"